### DISCUSSION TITLE
Sign in with email throttling. possible?

### DISCUSSION THREAD

**Discussion by simonwidjaja** [nextauthjs/next-auth#687] [Help]
**Your question**
I am a huge fan of next-auth so far! But I just discovered that I can click on Signin with Email several times. And for each click an email will be sent to the provided Email. One should be enough. Otherwise I could spam someone else's email account easily. Or am I missing something? 

IMHO: Once there is an entry in verificationRequests (with expires time in the future) there should not be sent another email.

Your thoughts?

**Comment by s-kris** [comment]
I think throttling is too subjective to implement in core-library.
You can implement functionality manually by checking for previous verificationRequests and then calling signIn function.

**Comment by simonwidjaja** [comment]
Is it really subjective? I think it's pretty critical from a GDRP
perspective. With the default settings I was able to send 10+ emails in 1s
to a 3rd party email account. I was finding this out by accident. I was
trying to sign in with email and the server was slow. So I clicked the
"Sign In" button several times. A second later my iPhone was vibrating like
crazy and my email inbox was pretty full... But that are just my 2 cents...

On Sat, Sep 19, 2020 at 11:30 AM Sai Krishna <notifications@github.com>
wrote:

> I think throttling is too subjective to implement in core-library.
> You can implement functionality manually by checking for previous
> verificationRequests and then calling signIn function.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nextauthjs/next-auth/issues/687#issuecomment-695190480>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANNBVVG34OBCBZHOTKT4N3SGR24LANCNFSM4RRWZJ2A>
> .
>


-- 

--
Simon Widjaja
CEO
Web professional driven by pure passion
Adobe Certified Expert (ACE)
Adobe Certified Instructor (ACI)
Adobe Community Professional (ACP)
Author and trainer for interactive media
Frequent speaker at int. conferences (Adobe MAX, FITC, Reasons…)

**Comment by iaincollins** [comment]
@simonwidjaja Hi Simon!

I think this a reasonable feature request and we should have some form of rate limiting on providers like email that use the verification request type. We should probably have a sensible rate limit in place by default.

We could also improve the behaviour of the front end default sign in page behaviour if that's a cause of triggering multiple emails during sign in (e.g. if users are clicking multiple times to sign in because the backend SMTP service is slow).

I also agree we should encourage best practice around opt out too, perhaps by encouraging people to provide an opt-out URL or showing how to use an SMTP service that offers this option (e.g. by appending an opt out URL to the bottom of each message).

**Comment by stale** [comment]
Hi there! It looks like this issue hasn't had any activity for a while. It will be closed if no further activity occurs. If you think your issue is still relevant, feel free to comment on it to keep ot open. Thanks!

**Comment by s-kris** [comment]
@balazsorban44 I came back to this! Any plans to implement this?

**Reply by balazsorban44** [reply]
Would be nice to open this as a feature request for better visibility. If we get the details done and maybe someone from the community have the time to implement this, I don't see a problem it being implemented. Currently I have my focus elsewhere (getting the adapters repo ready for stable, and adding tests to core is the most important for me right now.)

**Reply by s-kris** [reply]
Sure, that sounds good. I can take it up!

**Reply by 4nduril** [reply]
Hi, just stumbled upon the same question. I guess there is no further work done on this topic yet.
For me it's not clear where and with which tools to build a rate limit myself. Any hints how I could start?

**Comment by FilipGrebowski** [comment]
Has this issue been addressed already?

**Comment by Erel0** [comment]
Upstash have rate limiting SDK. I think we can handle this issue with limiting '/api/auth/signin/email' , but we have to request endpoint serverside.

- https://docs.upstash.com/redis
- https://github.com/upstash/ratelimit

**Comment by YannHulot** [comment]
For anyone who wonders how to do that.

Basically, you need to create an endpoint to send the magic link manually. In this endpoint, you need to use your rate limiter. 

You ll also need to duplicate some logic from the `signIn` method(from the client side) to generate the correct callback url and the proper verification token.

If everything is done properly, you can now rate limit the email sign up flow and your users should be able to login with their emails.

**Reply by YannHulot** [reply]
I know this is an old discussion but this is where I ended when I went looking for a solution and I couldn't find one. So here you go.
